File: wampiriada-lib/src/ImageGrid/TileSequence.php
<?php namespace NZS\Wampiriada\ImageGrid;

class TileSequence {
    public function __construct($tiles, $seed) {
        $this->tiles = $tiles;
        $this->shuffleTiles($seed);
    }
    private function shuffleTiles($seed) {
        $medianWeight = $this->getMedianWeight();
        foreach ($this->tiles as $key => $value) {
            $this->tiles[$key]['rare'] = ($value['weight'] > $medianWeight);
        }
        $nextSeed = rand();
        srand($seed); // Entering deterministic zone
        shuffle($this->tiles);
        $j = count($this->tiles) - 1;
        for ($i=0; $i<$j; ++$i) {
            if ($this->tiles[$i]['rare']) {
                $k = rand($i+1, $j);
                $tmp = $this->tiles[$i];
                $this->tiles[$i] = $this->tiles[$k];
                $this->tiles[$k] = $tmp;
                $j -= 1;
            }
        }
        srand($nextSeed); // Exiting deterministic zone
    }

    private function getMedianWeight() {
        usort($this->tiles, function($a, $b) {
            return $b['weight'] - $a['weight'];
        });
        return $this->tiles[floor(count($this->tiles)/2)]['weight'];
    }

    public function next() {
        return array_shift($this->tiles);
    }
}

File: image_grid/tile_sequence.py
from random import Random


class TileSequence(object):
    def __init__(self, tiles, seed):
        self.tiles = tiles
        self.shuffleTiles(seed)

    def shuffleTiles(self, seed):
        medianWeight = self.getMedianWeight()
        for tile in self.tiles:
            tile['rare'] = (tile['weight'] > medianWeight)

        if seed:
            rng = Random(seed)
        else:
            rng = Random()
        rng.shuffle(self.tiles)
        i = 0
        j = len(self.tiles) - 1
        while i < j:
            if self.tiles[i]['rare']:
                k = rng.randint(i+1, j)
                tmp = self.tiles[i]
                self.tiles[i] = self.tiles[k]
                self.tiles[k] = tmp
                j -= 1
            i += 1

    def getMedianWeight(self):
        self.tiles.sort(key=lambda tile: tile['weight'])
        return self.tiles[len(self.tiles) // 2]['weight']

    def __iter__(self):
        return iter(self.tiles)
